### PR TITLE
[SPARK-38688][SQL][TESTS] Use error classes in the compilation errors of deserializer

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
Migrate the following errors in QueryCompilationErrors:
 
* dataTypeMismatchForDeserializerError -> UNSUPPORTED_DESERIALIZER.DATA_TYPE_MISMATCH
* fieldNumberMismatchForDeserializerError -> UNSUPPORTED_DESERIALIZER.FIELD_NUMBER_MISMATCH

### Why are the changes needed?
Porting compilation errors of unsupported deserializer to new error framework.
 
### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add new UT.

### CODE REVIEW

**Comment by panbingkun** [comment]
pinging @MaxGekk

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 1]
```diff
@@ -147,14 +147,17 @@ object QueryCompilationErrors extends QueryErrorsBase {
       dataType: DataType, desiredType: String): Throwable = {
     val quantifier = if (desiredType.equals("array")) "an" else "a"
     new AnalysisException(
-      s"need $quantifier $desiredType field but got " + dataType.catalogString)
+      errorClass = "UNSUPPORTED_DESERIALIZER",
+      messageParameters = Array("DATA_TYPE_MISMATCH", quantifier, desiredType,
+        dataType.catalogString))
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:152]: Please, use toSQLType() to quote types.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 2]
```diff
@@ -147,14 +147,17 @@ object QueryCompilationErrors extends QueryErrorsBase {
...
   }
 
   def fieldNumberMismatchForDeserializerError(
       schema: StructType, maxOrdinal: Int): Throwable = {
     new AnalysisException(
-      s"Try to map ${schema.catalogString} to Tuple${maxOrdinal + 1}, " +
-        "but failed as the number of fields does not line up.")
+      errorClass = "UNSUPPORTED_DESERIALIZER",
+      messageParameters = Array("FIELD_NUMBER_MISMATCH", schema.catalogString,
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:159]: Quote the schema by toSQLType()

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 3]
```diff
@@ -147,14 +147,17 @@ object QueryCompilationErrors extends QueryErrorsBase {
       dataType: DataType, desiredType: String): Throwable = {
     val quantifier = if (desiredType.equals("array")) "an" else "a"
     new AnalysisException(
-      s"need $quantifier $desiredType field but got " + dataType.catalogString)
+      errorClass = "UNSUPPORTED_DESERIALIZER",
+      messageParameters =
+        Array("DATA_TYPE_MISMATCH", quantifier, desiredType, toSQLType(dataType)))
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:152]: Please, double quote `desiredType` and upper case it for consistency.

**Review by MaxGekk — COMMENTED** [review]
LGTM except of a comment.

**Review by MaxGekk — APPROVED** [review]
+1, LGTM. Merging to master.
Thank you, @panbingkun.

**Comment by MaxGekk** [comment]
@panbingkun Since this PR modified error classes, could you backport it to branch-3.3, please.

**Pull request closed without merging**